openverse_catalog/dags/data_refresh/dag_factory.py

```diff
@@ -0,0 +1,186 @@
+    """
+    This factory method instantiates a DAG that will run the data refresh for
+    the given `media_type`.
+
+    A data refresh runs for a given media type in the API DB. It imports the
+    data for that type from the upstream DB in the Catalog, reindexes the data,
+    and updates and reindex Elasticsearch.
+
+    A data refresh can only be performed for one media type at a time, so the DAG
+    must also use a Sensor to make sure that no two data refresh tasks run
+    concurrently.
+
+    It is intended that the data_refresh tasks, or at least the initial
+    `wait_for_data_refresh` tasks, should be run in a custom pool with 1 worker
+    slot. This enforces that no two `wait_for_data_refresh` tasks can start
+    concurrently and enter a race condition.
+
+    Required Arguments:
+
+    data_refresh:     dataclass containing configuration information for the
+                      DAG
+    external_dag_ids: list of ids of the other data refresh DAGs. This DAG
+                      will not run concurrently with any dependent DAG.
+    """
+    dag = DAG(
+        dag_id=data_refresh.dag_id,
+        default_args=data_refresh.default_args,
+        start_date=data_refresh.start_date,
+        schedule_interval=data_refresh.schedule_string,
+        catchup=False,
+        doc_md=data_refresh.doc_md,
+        tags=[f"data_refresh: {data_refresh.media_type}"],
+    )
+
+    with dag:
+        # Wait to ensure that no other Data Refresh DAGs are running.
+        wait_for_data_refresh = ExternalDAGsSensor(
+            task_id="wait_for_data_refresh",
+            external_dag_ids=external_dag_ids,
+            check_existence=True,
+            pool=DATA_REFRESH_POOL,
+            dag=dag,
+            poke_interval=5,
+            mode="reschedule",
+        )
+
+        data_refresh_post_data = {
+            "model": data_refresh.media_type,
+            "action": "INGEST_UPSTREAM",
+        }
+
+        # Trigger the refresh on the data refresh server.
+        trigger_data_refresh = SimpleHttpOperator(
+            task_id="trigger_data_refresh",
+            http_conn_id="data_refresh",
+            endpoint="task",
+            method="POST",
+            headers={"Content-Type": "application/json"},
+            data=json.dumps(data_refresh_post_data),
+            response_check=lambda response: response.status_code == 202,
+            response_filter=response_filter_data_refresh,
+            pool=DATA_REFRESH_POOL,
+            dag=dag,
+        )
+
+        # Wait for the data refresh to complete.
+        wait_for_completion = HttpSensor(
+            task_id="wait_for_completion",
+            http_conn_id="data_refresh",
+            endpoint=XCOM_PULL_TEMPLATE.format(
+                trigger_data_refresh.task_id, "return_value"
+            ),
+            method="GET",
+            response_check=response_check_wait_for_completion,
+            pool=DATA_REFRESH_POOL,
```

Comment:
Since `pool` is used in every task in this DAG, we can actually just add it to the DAG's `default_args`! That will apply it automatically to every task.

Reply:
Nice! I _think_ it's actually only important that the `wait` task be in the data refresh pool, actually, since that should gate the other tasks. But I think I'll keep it on all of them as it's a little easier to reason about. I'll add this to clean it up 🎉

Reply:
I agree, it should only be necessary there. But the actual task check times aren't too long, so I think having them all share the pool is fine!